nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+      throw new Error('.register() did not return a promise when it should have.');
+    }
+    const registrationResult = await registrationResultPromise.catch(e => {throw new Error('DPS registration failed: ' + e.message)});
+    const connectionString = 'HostName=' + registrationResult.assignedHub + ';DeviceId=' + registrationResult.deviceId + ';SharedAccessKey=' + deviceKey;
+    const hubClient = Client.fromConnectionString(connectionString, IotHubTransport);
+    await hubClient.open().catch(e => {throw new Error('Opening client failed: ' + e.message)});
+    return hubClient;
+  }
+
+  private pairWithService() {
+    console.info('Starting pairing operation.');
+    const pairingRequestPatch = {
+      thief: {
+        pairing: {
+          requestedServicePool: requestedServicePool,
+          serviceInstanceId: null,
+          runId: runId
+        }
+      }
+    };
+
+    const pairWithServiceAttemptRecursive = () => {
+      return new Promise<void>((res, rej) => {
+        console.info('Updating pairing reported props: %j', pairingRequestPatch);
+        promisify(this.twin.properties.reported.update)(pairingRequestPatch)
+          .then(() => {
+            // Register an event listener to listen for response from service app
+            this.twin.on('properties.desired.thief.pairing', received => {
+              if (this.serviceInstanceId) {
+                console.info('Already paired. Ignoring');
+              }
+              else if (!received.runId || !received.serviceInstanceId) {
+               console.info('runId and/or serviceId is missing. Ignoring.');
+              }
+              else if (received.runId !== runId) {
+                console.info(`runId mismatch. Ignoring. (received ${received.runId}, expected ${runId})`);
+              }
+              else {
+                console.info(`Service app ${received.serviceInstanceId} claimed this device instance`);
+                this.twin.removeAllListeners('properties.desired.thief.pairing');
+                clearTimeout(this.pairingAttemptTimer);
+                this.serviceInstanceId = received.serviceInstanceId;
+                const pairingAcceptPatch = {
+                  thief: {
+                    pairing: {
+                      serviceInstanceId: received.serviceInstanceId,
+                      runId: runId
+                    }
+                  }
+                };
+                console.info('Updating pairing reported props: %j', pairingAcceptPatch);
+                promisify(this.twin.properties.reported.update)(pairingAcceptPatch)
+                  .then(res)
+                  .catch(e => rej(new Error('Updating reported properties failed: ' + e.message)));
+              }
+            });
+
+            // If it's taking too long for a service app to respond to our pairing request,
+            // retry by recursively calling pairWithServiceAttemptRecursive()
+            this.pairingAttemptTimer = setTimeout(() => {
```

Comment:
there's a nice way to create a promisified setTimeout since all your code is already async await. Create a `wait` method that is just 

```js
function wait (time) {
    new Promise((resolve) => {
        setTimeout(resolve, time);
    })
}
```

Reply:
eh I take this back, for your use cases this is easier to comprehend.